### DISCUSSION TITLE
Blazor generating single session id for user session

### DISCUSSION THREAD

**Discussion by SethSikoia** [dotnet/aspnetcore#51704] [General]
I have a blazor server side application and I want to implement some logging logic to track user activity. I am using structured logging and I want to add a property that I can use to filter all logs related to a given user session. 

The word session can have a different meaning given the context, what I mean in this case by session is one visit of a user to the application, so given then user opens a new tab that will start a new session.  

We have created a service that will handle generating the session id and adding it as a property in the logger using Logger.BeginScope, based on our tests, none of the available service registration options seems to be fit for purpose. Singleton service seems to be global for the whole application so all user share the same session id, scoped service seemed to be the one that would work as supposedly there is one for each blazor circuit. However the behaviour exhibited in practice is that the scoped application service is instantiated more than once in the same user session, meaning that multiple session id's are produced. 

           
```
services.AddScoped<SessionService>();

public class SessionService
{
    public string SessionId { get; }

    public SessionService()
    {
        SessionId = Guid.NewGuid().ToString();
    }
}

<h3>Foo</h3>

@code {

    @inject SessionService SessionService {get;set;}
    @inject Logger<Foo> Logger {get;set;}

    protected override Task OnInitializedAsync()
    {
        var sessionId = SessionService.SessionId; 
        Logger.LogInformation($"Session Id: {sessionId}");

        return base.OnInitializedAsync();
    }


}
```

What is the best way to guarantee the session Id will be consistent for the lifetime of the users session?

**Comment by guy-avery** [comment]
I think the answer you want is to register the class in the Di as Scoped, not as a singleton. Scoped instances are directly tied to the session so only live as long as the user's session. Just be aware that doing things like a page refresh will normally create a new session, so the same user would end up with a new Id in this scenario. There is a full example of what you are trying to do here:
https://stackoverflow.com/questions/72269666/how-to-store-an-unique-blazor-sessionid-for-each-client
